### ISSUE TITLE
Proxy settings not configured for all clients

### ISSUE THREAD

**Issue by corrad82-4s** [DSpace/DSpace#8603]
**Describe the bug**
We have found some connector / classes making calls to external systems where HttpClient is initialized without proxy settings. For example in `OrcidV3AuthorDataProvider` class, client is initialized without proxy settings (https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV3AuthorDataProvider.java#L90) 

**Expected behavior**
All used HttpClient instance should take into account and set proxy settings. Could a centralized / utility class responsible to provide an instance of HttpClient be worth? 
While checking this behavior, we have seen that proxy parameters are set via DSpace configuration, while it could be also possible to define proxy parameters at VM level (https://www.baeldung.com/java-connect-via-proxy-server , https://stackoverflow.com/questions/6780128/why-dont-proxyhost-proxyport-work-when-running-my-java-application - see 2nd answer). Shall we keep proxy configuration via DSpace configuration files or it could be worth moving them from configuration to VM?

**Comment by arvoConsultores** [comment]
currently working on this issue.

**Comment by tdonohue** [comment]
@sergius02 : For what it's worth, I'm not aware of any reason why some classes use `HttpClient` versus `URLConnection`.  I suspect it may have just been by accident (the original developer chose what they were familiar with). So, I agree with you. If possible, it'd be nice to unify all those calls under a single utility class.

**Comment by arvoConsultores** [comment]
Classes  using some of the aforementioned methods

- Sherpa (SHERPAService.java)
- Orcid (OrcidRestConnector.java)
- Orcidv3 (Orcidv3SolrAuthorityImpl.java, OrcidV3AuthorDataProvider.java)
- MicrosoftTranslator (MicrosoftTranslator.java)
- GoogleCaptcha (CaptchaServiceImpl.java)
- OpenAIRE (OpenAIRERestConnector.java)
- Google Analytics (GoogleRecorderEventListener.java, GoogleAnalyticsClientImpl.java)
- DataCite (DataCiteConnector.java)
- EZID (EZIDRequest.java)
- CC license (CCLicenseConnectorServiceImpl.java) < in this one both connection methods are used¡¡

**Comment by arvoConsultores** [comment]
and  LiveImportClientImpl class   that is used by all the item-import APIs
Crossref, Pubmed, ADS, Cinii, DataCite, Epo, Pubmed europe, Scielo, Scopus, VuFind, WOS

**Comment by ckubgi** [comment]
Hello. I found this issue while looking for a solution on how to set a proxy for all requests to external resources. We are required by our network admins to do it that way. 
Do I understand correctly that there is currently no way to make DSpace use a proxy for all requests to external resources?

**Comment by tdonohue** [comment]
@ckubgi : No, that's incorrect.  DSpace will use a configure proxy for requests to external resources. But there are a few newer external resource plugins which are NOT correctly using those proxy settings -- namely ORCID integration.  So, if you are using ORCID integration, that is not working with a proxy. But, if you are not using ORCID integration, then all other activities should work with a proxy (as far as I'm aware).  See DSpace's proxy settings at https://github.com/DSpace/DSpace/blob/main/dspace/config/dspace.cfg#L404-L440

**Comment by ckubgi** [comment]
@tdonohue after having a hard time trying to configure the proxy, let me summarize my experience so far:

Our staging system is behind a proxy. 
When I configured dspace's proxy settings as the documentation suggests, essential external services like CC-License and DOI are getting connection timeouts.
Checking outgoing connections showed me that these services try to connect directly and ignore my proxy settings.
That puzzled me, and I started to look into the code and found that these proxy settings are used by `CreativeCommonsServiceImpl.java` here:
https://github.com/DSpace/DSpace/blob/f072e9af3d051e0953973679db9171e781e21437/dspace-api/src/main/java/org/dspace/license/CreativeCommonsServiceImpl.java#L108-L115

After those properties are set, the actual connection is handled by `CCLicenseConnectorServiceImpl.java`
which builds an `HttpClient` with no explicit configuration of a proxy.
https://github.com/DSpace/DSpace/blob/f072e9af3d051e0953973679db9171e781e21437/dspace-api/src/main/java/org/dspace/license/CCLicenseConnectorServiceImpl.java#L73-L78

So i assumed, it uses system properties implicitly because otherwise those configuration settings would be useless, at least to `CCLicenseConnectorServiceImpl.java`

I then thought it simply missed a configuration for https because my examples, DOI and CC-License are requested via https. 
As far as i understand it, using `System.setProperty` has the same or a similar effect
as setting `JAVA_OPTS="${JAVA_OPTS} -Dhttp.proxyHost=<proxy_hostname> -Dhttp.proxyPort=<port_number>"`
in my `/etc/default/tomcat9` file (Ubuntu 22.04).
I added `-Dhttps.proxyHost=<proxy_hostname> -Dhttps.proxyPort=<port_number>` and expected these values to be available to dspace.

Still ignored.
Searching around, I found a useful StackOverflow thread:
https://stackoverflow.com/questions/6780128/why-dont-proxyhost-proxyport-work-when-running-my-java-application

And I found a tutorial on how to setup a proxy for Apache HttpClient here:
https://www.baeldung.com/httpclient-advanced-config

Both sources and my tests indicated to me that at least in `CCLicenseConnectorServiceImpl.java`, the HttpClient is not created in a way that respects the system properties, and it's also not possible to set a value for https.proxHost in dspace.cfg.

This was quite a rabbit hole for me, so I might get something wrong here.
I would really appreciate some help. If this is not the right place to discuss this, please let me know where it would be more appropriate.

**Comment by tdonohue** [comment]
@ckubgi : I suspect you are seeing the same thing that is reported in the comment above by @arvoConsultores (who is assigned to this ticket). https://github.com/DSpace/DSpace/issues/8603#issuecomment-1662674742

In that comment, they mentioned:
```
CC license (CCLicenseConnectorServiceImpl.java) < in this one both connection methods are used!!
```

So, it sounds like that particular connector may have issues as well since it is using two different connection types, and only setting the proxy in one.  

Again though, the DSpace Proxy settings *work* for some clients/connectors, but we are aware they do *not* work for all clients/connectors.  The purpose of this ticket is to fix all these clients to work in the same way, using a centralized proxy utility class.  So, I suspect you may need to wait on this ticket to be resolved before it will work fully...I'm not sure what the status of the fixes is at this time, but @arvoConsultores mentioned they are "working on it" as of August.  

So, I believe your report is simply verifying to us that the CC License service is definitely one of those that is broken.

I'm moving this to our 7.6.x maintenance board to see if we can get this moving forward for a bug-fix release.  @arvoConsultores please do let us know the latest status to see if we can get this fixed sooner rather than later.

**Comment by mwoodiupui** [comment]
I needed to make a patch for CC and found this.  I will be happy to cooperate with anyone else working on pieces of this puzzle.

**Comment by kshepherd** [comment]
fwiw, as part of some recent work that involved a lot of external source integration, I made what I think is quite a decent `AbstractRestConnector` which handles some low level stuff, including the ability to easily mock responses from disk for integration tests, and could be a way for us to plug in proxy-awareness for a lot of REST connectors like ORCID, SHERPA, OpenAire, etc.
If there is interest in this I will extract just the REST connector impl from my project and make a PR (I have plans to chop it up and contribute it back in several pieces over time anyway!)
